includes/printf.h

```diff
@@ -0,0 +1,14 @@
+#ifndef PRINTF_H
+#define PRINTF_H
+
+#include <stddef.h>
+
+typedef struct  writer_s
+{
+    void (*write)(struct writer_s *self, const char buf[], size_t len);
+    void *data;
+}               writer;
+
+void printf(writer *wrt, const char *format, ...);
```

Comment:
nit. This is now `fprintf`. You could create two functions: `fprintf` that takes in the writer, and `printf` that uses some default writer. This would allow you to easily switch the default writer in one place